### PR TITLE
MRTK Profile Inspector Performance Improvements

### PR DESCRIPTION
**Pull request by Railboy**
## Overview
Major changes
- Adds the `MixedRealityProfileUtility` class. This class is responsible for listening for asset changes and caching available profiles by type and service type. Prior to this change inspectors were searching for ScriptableObject instances every redraw.
- Changes profiles from object fields to dropdown lists and moves profile drawing from `BaseMixedRealityProfileInspector` to `MixedRealityInspectorUtilities`. The GUIContent for these lists are cached.
- Adds a 'View Asset' button to profile fields - clicking this reveals the ScriptableObject asset in the project view.
- Removes object picker window functionality from MixedRealityToolkitInspector, since dropdowns are effectively the same.

MRTKInspector

## Changes
- Addresses: #6120

### CODE REVIEW

File: Assets/MixedRealityToolkit/Extensions/EditorClassExtensions/ScriptableObjectExtensions.cs [hunk 1]
```diff
@@ -67,5 +67,25 @@ public static ScriptableObject CreateAsset(this ScriptableObject scriptableObjec
 
             return instances;
         }
+
+        /// <summary>
+        /// Gets all the scriptable object instances in the project.
+        /// </summary>
+        /// <param name="profileType">The Type of <see href="https://docs.unity3d.com/ScriptReference/ScriptableObject.html">ScriptableObject</see> you're wanting to find instances of.</typeparam></param>
+        /// <returns></returns>
+        public static ScriptableObject[] GetAllInstances(Type profileType)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Extensions/EditorClassExtensions/ScriptableObjectExtensions.cs]: add where T : BaseMixedRealityProfile (or whatever the class)

**Comment by julenka** [comment]
Seems related to #5791, did you know about that PR? Can you share similarities and differences?

File: Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityInspectorUtility.cs [hunk 2]
```diff
@@ -453,5 +454,130 @@ public static Quaternion RotationHandle(Object target, Vector3 position, Quatern
...
+
+            var subProfile = profileObject as BaseMixedRealityProfile;
+            if (subProfile != null && !subProfile.IsCustomProfile)
+            {
+                EditorGUILayout.HelpBox("Clone this default profile to edit properties below", MessageType.Warning);
+            }
+
+            if (renderProfileInBox)
+            {
+                EditorGUILayout.BeginVertical(EditorStyles.helpBox);
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityInspectorUtility.cs]: switch maybe to switch to new EditorGUILayout.VerticalScope(EditorStyles.HelpBox)
and use ? : operator to pass correct style in?

**Comment by julenka** [comment]
I really like the dropdown idea! Will make profiles much easier to find.

File: Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+            }
+            return false;
+        }
+
+        /// <summary>
+        /// Returns true if profile is NOT a BaseMixedRealityProfile class type.
+        /// </summary>
+        /// <param name="profileTypeName"></param>
+        /// <returns></returns>
+        public static bool IsConcreteProfileType(String profileTypeName)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs]: maybe this should use the Type class instead of doing strings, seems like could easily break..especially if class names change etc

File: Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs [hunk 4]
```diff
@@ -50,64 +45,14 @@ public override void OnInspectorGUI()
...
-            EditorGUILayout.PropertyField(activeProfile);
-            bool changed = EditorGUI.EndChangeCheck();
-            string commandName = Event.current.commandName;
 
             // If not profile is assigned, then warn user
             if (activeProfile.objectReferenceValue == null)
             {
                 EditorGUILayout.HelpBox("MixedRealityToolkit cannot initialize unless an Active Profile is assigned!", MessageType.Error);
-
-                if (GUILayout.Button("Assign MixedRealityToolkit Profile") || forceShowProfilePicker)
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs]: Why is this deleted?

**Comment by Troy-Ferrell** [comment]
Reading the issue you filed, I saw you said "Observer poor performance", this is more out of curiosity but what are you seeing? I actually haven't noticed any issues myself (but of course different projects/dev machines etc etc). If you run a scene and show the inspector on the side, do you notice a before/after in FPS on the visual profiler

File: Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs (hunk 4 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit/Inspectors/MixedRealityToolkitInspector.cs]: The profile picker is no longer necessary now that we have dropdown menus.

File: Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs (hunk 3 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs]: Probably a good idea, but my goal was to mess with existing functionality as little as possible. I'd want to handle this in a separate PR that was less performance-focused.

**Comment by Railboy** [comment]
>Reading the issue you filed, I saw you said "Observer poor performance", this is more out of curiosity but what are you seeing?

@Troy-Ferrell maybe @Cameron-Micka can share some of the feedback we've been getting. In my experience button clicks are sluggish and resizing the inspector is tough, especially in 2019.

>Out of curiosity, do you have numbers on the perf difference before/after?

I don't - it's tricky to guage because it varies based on how many / which profile inspectors are rendering at once. But each call to ScriptableObjectExtensions was taking anywhere from 20-120ms, and many happening each OnInspectorGUI call, so you could get some pretty laggy spikes. With this change you only see those spikes the first time a profile type is rendered.

**Comment by Railboy** [comment]
>Seems related to #5791, did you know about that PR? Can you share similarities and differences?

@julenka We're trying to solve the same problem but we're caching things in different ways. This PR responds better to asset changes (not just recompilation) so the cache doesn't get stale.

**Comment by Cameron-Micka** [comment]
@Troy-Ferrell our partner onsite this week was seeing (if I had to guesstimate) about a 5Hz refresh of the MRTK Profile inspector when using it. Resizing the inspector it became very apparent. Their laptops were pretty modest, and they were using Unity 2019. So maybe there is a 2019 regression?

File: Assets/MixedRealityToolkit/Extensions/EditorClassExtensions/ScriptableObjectExtensions.cs (hunk 1 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit/Extensions/EditorClassExtensions/ScriptableObjectExtensions.cs]: This isn't a generic method.

**Comment by Railboy** [comment]
@Troy-Ferrell If they are the same, then why change?

They're the same in the sense that they both present a set of available options. But I feel dropdowns are more user-friendly. A dropdown removes all ambiguity about what to select and how. An object field + a separate pop-up window is an opportunity for accidental closure / confusion about its relationship to the profile / etc. I feel the same way about the clone window - someday I'm going to get rid of that thing.

File: Assets/MixedRealityToolkit/Extensions/EditorClassExtensions/ScriptableObjectExtensions.cs (hunk 1 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Extensions/EditorClassExtensions/ScriptableObjectExtensions.cs]: That works but then rename argument to be assetType or typeFilter etc not profile type

**Comment by Troy-Ferrell** [comment]
> @Troy-Ferrell If they are the same, then why change?
> They're the same in the sense that they both present a set of available options. But I feel dropdowns are more user-friendly. A dropdown removes all ambiguity about what to select and how. An object field + a separate pop-up window is an opportunity for accidental closure / confusion about its relationship to the profile / etc. I feel the same way about the clone window - someday I'm going to get rid of that thing.

That's true. I hate when the popup window goes to my far left monitor but I am working on my far right, etc. 

I just hate dropdowns when I am searching for something. # of profiles should be pretty small, especially since the results are filtered by type. it annoys the hell out of me when doing something like shader properties.

File: Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs (hunk 3 above)
Comment (Troy-Ferrell) [reply Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs]: fair enough, but can we at least change it to use types while here. It doesn't need the additional if type.isclass or type.isabstract etc
public static bool IsConcreteProfileType(Type profileType)
{
return profileType != typeof(BaseMRClass);
}

File: Assets/MixedRealityToolkit/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs [hunk 5]
```diff
@@ -132,76 +130,13 @@ protected static bool RenderProfile(SerializedProperty property, Type profileTyp
...
-
-                if (renderProfileInBox)
-                {
-                    EditorGUILayout.BeginVertical(EditorStyles.helpBox);
-                }
-                else
-                {
-                    EditorGUILayout.BeginVertical();
-                }
+            Debug.Assert(profile != null, "No profile was set in OnEnable. Did you forget to call base.OnEnable in a derived profile class?");
```
Comment (Troy-Ferrell) [inline Assets/MixedRealityToolkit/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs]: Why would profile be set in OnEnable? Not understanding the assert comment?

Also should this render a warning/error in the inspector? Because this will likely spam the console no? On every repaint

File: Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+        }
+
+        private static void AssetImportListener_OnAssetsChange()
+        {
+            RefreshProfileCaches();
+        }
+
+        private static void RefreshProfileCaches()
+        {
+            List<Type> cachedTypes = new List<Type>(profileCaches.Keys);
```
[the remark this answers is not in the document]
Comment (Railboy) [reply Assets/MixedRealityToolkit/Inspectors/Utilities/MixedRealityProfileUtility.cs]: People are expecting a lengthy load time on asset import / recompile so we're using that as an opportunity to do some painful work.

File: Assets/MixedRealityToolkit/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs (hunk 5 above)
Comment (Railboy) [reply Assets/MixedRealityToolkit/Inspectors/Profiles/BaseMixedRealityProfileInspector.cs]: It was a check in the original base class. I've never actually seen this assert fail in the wild. I'm going to remove it.

**Comment by Troy-Ferrell** [comment]
@Railboy MixedRealityProfileUtility has doccomments with empty returns/params
